state/watcher.go

```diff
@@ -1081,20 +1089,57 @@ func setRelationUnitChangeVersion(changes *params.RelationUnitsChange, key strin
 	changes.Changed[name] = settings
 }
 
+func (w *relationUnitsWatcher) watchRelatedAppSettings(changes *params.RelationUnitsChange) error {
+	idsAsInterface := make([]interface{}, len(w.appSettingsKeys))
+	for i, key := range w.appSettingsKeys {
+		idsAsInterface[i] = w.backend.docID(key)
+	}
+	w.logger.Tracef("relationUnitsWatcher %q watching app keys: %v", w.sw.prefix, w.appSettingsKeys)
+	if err := w.watcher.WatchMulti(settingsC, idsAsInterface, w.appUpdates); err != nil {
+		return errors.Trace(err)
+	}
```

Comment:
I hit a bug where watchers wouldn't fire after upgrading the model because the finish method doesn't unwatch these docs - adding them to .watching resolved it.
```
	for _, docID := range idsAsInterface {
		w.watching.Add(docID.(string))
	}
```